c10/core/DispatchKey.h

```diff
@@ -661,37 +709,50 @@ constexpr BackendComponent toBackendComponent(DispatchKey k) {
   if (k >= DispatchKey::StartOfDenseBackends &&
       k <= DispatchKey::EndOfDenseBackends) {
     return static_cast<BackendComponent>(
-        static_cast<uint8_t>(k) -
-        static_cast<uint8_t>(DispatchKey::StartOfDenseBackends));
+        static_cast<uint16_t>(k) -
+        static_cast<uint16_t>(DispatchKey::StartOfDenseBackends));
```

Comment:
oh this was subtle - a bunch of these casts need to be `uint16_t` now, since the `DispatchKey` enum has more than 64 entries.

I'm not actually sure why this wasn't a problem before - it was caught now because of a `constexpr DispatchKeySet(DispatchKey::AutocastXPU)` call that tripped the compiler.

Reply:
Hmm, still looking into the CI errors. I haven't been able to repro them in godbolt

Reply:
I think we should probably start distinguishing between the bitmask (which only has 64 entries) and the extended cross-product set (which has a lot more entries)... maybe it would be better not to list them all out explicitly, and instead have some bitmasking to construct the combined index.

Reply:
Can we reasonably do that though without breaking people who are using the existing dispatch key enum vals from the cross product? e.g. `DispatchKey::AutocastXPU` or `DispatchKey::AutogradXLA`

Reply:
Oh yeah, it'd be BC-breaking. But it's just XLA it's not a big deal, we can fix them up.

Reply:
What about all of the other backend keys 😃 

(this includes the "dense" keys too, since they're technically not part of the 64-bit range, like `ORT`, `XPU`, `IPU`, etc)

Reply:
Well, another way to make it BC is to grandfather in the old keys (but have a new style that doesn't require writing combinatorially many keys). Still seems better than maintaining huge tables.

Reply:
Returning to this, we should consider refining the types, so we have one type for what can go in the bitset, and another type for the extended set. In any case, I don't think we are relying on the size of the dispatch key in a strict (since they go in bitset) so increasing the size seems fine.